### ISSUE TITLE
Consensus hash divergence with newer boost (>=1.65.0)

### ISSUE THREAD

**Issue by dgpv** [OmniLayer/omnicore#872]
There is a bug that causes checkpoint failure at block 310000 (https://github.com/mpugach/docker_omnicored/issues/2).

It happens because:

https://github.com/OmniLayer/omnicore/blob/c0738a96b587c80a73f93cb653c032e1cc95b6f4/src/omnicore/log.cpp#L55

https://github.com/OmniLayer/omnicore/blob/c0738a96b587c80a73f93cb653c032e1cc95b6f4/src/omnicore/tx.cpp#L528

https://github.com/OmniLayer/omnicore/blob/c0738a96b587c80a73f93cb653c032e1cc95b6f4/src/omnicore/tx.cpp#L539

Certain deadline dates are not valid, like in tx 5c183d6c1c9b8a617dbdcd1afa876eec14b46a2aa45d471e8381b287ccba731d

With new versions of boost (>= 1.65.0), boost::posix_time::from_time_t() throws an exception on invalid datetime.

```
2019-01-26 16:38:09 parseTransaction(block=306906, 2014-06-20 22:34:22 idx= 705); txid: b7c66175a99ca0e7b1691905d50a46165adb7a8012d9ec5e1ecf8239f859df6d
2019-01-26 16:38:09     ------------------------------
2019-01-26 16:38:09              version: 1, class B
2019-01-26 16:38:09                 type: 51 (Create Property - Variable)
2019-01-26 16:38:09            ecosystem: 2
2019-01-26 16:38:09        property type: 2 (divisible)
2019-01-26 16:38:09     prev property id: 0
2019-01-26 16:38:09             category: Testing
2019-01-26 16:38:09          subcategory: Crowdsale Demo
2019-01-26 16:38:09                 name: OmpaLoompa
2019-01-26 16:38:09                  url: http://www.themulticurrencycrowdsale.com
2019-01-26 16:38:09                 data: This awesome crowdsale supports multiple currencies!
2019-01-26 16:38:09     property desired: 2 (TOMNI)
2019-01-26 16:38:09      tokens per unit: 0.00000100
2019-01-26 16:38:09 
************************
EXCEPTION: N5boost16exception_detail10clone_implINS0_19error_info_injectorINS_9gregorian8bad_yearEEEEE       
Year is out of valid range: 1400..10000       
bitcoin in ProcessMessages()       

```
deadline here would be printed on old boost as:

```
2019-01-26 18:05:05             deadline: 1940-01-02 19:46:56 (146bc793d2b0)
```

When the same exception is caught while processing local blocks, omni core exits.
(https://github.com/OmniLayer/omnicore/issues/788, https://github.com/OmniLayer/omnicore/issues/731)

But when the node gets the block and transactions from the peer, this exception happens
in the context of `ProcessMessages()`, and does not result in shutdown -
there are code in `ProcessMessages()` that catches exception, and then just does `PrintExceptionContinue(&e, "ProcessMessages()");`

The crowdsale `CROWDSALE id: 2147483679 value: 100 property: 2`

was not registered due to this exception, but the processing continued.

Later, in tx 938a160cdb20c4103d2e5fc4922bd85be701797f41e64326f652469b7c8006d3, this causes

    2019-01-26 16:41:07 logicMath_SimpleSend(): rejected: sender 1KF9yZ2gnwgR4mu6d3ZA6yLJARhhGfBope has insufficient balance of property 2147483679 [0.00000000 < 8.58757932]

That results in consensus hash calculations to diverge from the clients that is running with old boost.

Because consensus hash diverged, checkpoint at block 310000 does not match, and omni core stops.

The danger here is that exceptions in consensus-critical code are ignored.

The checkpoint catched the bug, but the consensus hash diverged.

Here it happened with date formatting, but it may be other aspects of boost or other libs, and considering that some debug log flags are on by default, the code that outputs this debug logging messages might not consider that it might affect consensus hash.

Such mismatches can be discovered only after the node is restarted, and dies when it tries to process the data that was skipped when the exception was catched and ignored.

Are debug logging flags really necessary to be turned on by default ?

Should omni processing code catch own exceptions itself, and die on them ? That would turn impact of such bugs to denial of service, but not consensus divergence.

**Comment by dgpv** [comment]
I synced old-boost node past block 310000, and then run new-boost node with the same bitcoin data directory. New-boost node continued to sync, and also got exceptions in the log, that was ignored. This means that some of the nodes currently running may be out of sync with old-boost nodes (they will not see some omni transactions that is in the blocks on which there was exceptions caught), if these nodes are compiled with boost >= 1.65 (or, possibly, > 1.62 ?)

**Comment by dgpv** [comment]
The situation what I see were currently-running nodes might be at risk: Someone got new-boost version of Omni Core (built on new distro), but used old bitcoin data directory (created with old-boost node), to save time and not sync from scratch.

**Comment by dgpv** [comment]
After full sync with custom-patched omnicored (that catches the exceptions within DateTimeStrFormat and prints them, for the purpose of logging all the exceptions), as of block 561350, the last seen exception was in `parseTransaction(block=420551, 2016-07-13 11:57:14 idx= 451); txid: 5b81b20508c98a9bda3de68d6b3872f52efd02ebe2cb9b114cfb3a3225f33dd1`

This does not mean that such exceptions cannot happen in affected instances with future transactions.

All the exceptions catched since block 300000 to 561350: https://gist.github.com/dgpv/6b880893e4f310bbf81e3626ff94b027

**Comment by dgpv** [comment]
The last block that had the exception is block=420551 mined 2016-07-13

Boost version 1.65.0 is released 19 Aug 2017

So the probability of these old exceptions causing actual live consensus hash divergence is very low, because for this to happen, someone needs to get omnicored built after 19 Aug 2017 with boost 1.65.0, but use directory with blockchain data that was synced no later than block 420550, 10 months old directory.

Looking at omni core release notes, there should be a checkpoint at block 460000, which was added in Omni Core 0.2.0, so even that exception would be catched by the checkpoint.

This makes the probability of live consensus hash divergence even lower, because then it have to be pre-0.2.0 instance that got past all the checkpoints.

The potential danger is only in exceptions happening in new blocks.

**Comment by dexX7** [comment]
This was resolved with 0.4.0. Thank you so much again @dgpv!
